### PR TITLE
Push once

### PR DESCRIPTION
**Pull request by dgageot**
Image that were already pushed don't need to be pushed again

### CODE REVIEW

**Comment by dgageot** [comment]
Tests pass on my machine..

File: pkg/skaffold/docker/image.go [hunk 1]
```diff
@@ -218,6 +218,24 @@ func (l *localDaemon) Push(ctx context.Context, out io.Writer, ref string) (stri
 		return "", errors.Wrapf(err, "getting auth config for %s", ref)
 	}
 
+	// Check if the local image has a digest from a previous push
+	// If the remote image has the same digest, there's no need
+	// to push.
+	image, _, err := l.apiClient.ImageInspectWithRaw(ctx, ref)
+	if err == nil {
+		if len(image.RepoDigests) == 1 {
```
Comment (nkubala) [inline pkg/skaffold/docker/image.go]: i remember we had a brief conversation about this somewhere else....should this be `if len(image.RepoDigests) > 0`? i have no idea when or why there could be more than one digest but maybe we should account for that anyway?

Comment (dgageot) [reply pkg/skaffold/docker/image.go]: Yeah. My point was that I have no idea either. I decided to play it self.  I’ll just go through the normal path if there are many.

File: pkg/skaffold/docker/image.go [hunk 2]
```diff
@@ -218,6 +218,24 @@ func (l *localDaemon) Push(ctx context.Context, out io.Writer, ref string) (stri
 		return "", errors.Wrapf(err, "getting auth config for %s", ref)
 	}
 
+	// Check if the local image has a digest from a previous push
+	// If the remote image has the same digest, there's no need
+	// to push.
+	image, _, err := l.apiClient.ImageInspectWithRaw(ctx, ref)
+	if err == nil {
```
Comment (tejal29) [inline pkg/skaffold/docker/image.go]:
```suggestion
	if err == nil  && len(image.RepoDigests) == 1 {
```

File: pkg/skaffold/docker/image.go [hunk 3]
```diff
@@ -218,6 +218,24 @@ func (l *localDaemon) Push(ctx context.Context, out io.Writer, ref string) (stri
...
+		if len(image.RepoDigests) == 1 {
+			parsed, err := ParseReference(image.RepoDigests[0])
```
Comment (tejal29) [inline pkg/skaffold/docker/image.go]: refactor this as `isAlreadyPushed`. 
```suggestion
			if parsed, ok := isAlreadyPushed(image.RepoDigests[0]); ok {
			   return parsed.Digest, nil
			}
```

File: pkg/skaffold/docker/image_test.go [hunk 4]
```diff
@@ -76,6 +76,25 @@ func TestPush(t *testing.T) {
...
+		api.Add("image", "sha256:imageIDabcab")
+		localDocker := NewLocalDaemon(api, nil, false, nil)
+
+		digest, err := localDocker.Push(context.Background(), ioutil.Discard, "image")
+		t.CheckErrorAndDeepEqual(false, err, "sha256:bb1f952848763dd1f8fcf14231d7a4557775abf3c95e588561bc7a478c94e7e0", digest)
+
+		// Images already pushed don't need being pushed.
+		api.ErrImagePush = true
+
+		digest, err = localDocker.Push(context.Background(), ioutil.Discard, "image")
```
Comment (tejal29) [inline pkg/skaffold/docker/image_test.go]: test  `isAlreadyPushed` here instead.

File: pkg/skaffold/build/local/docker_test.go [hunk 5]
```diff
@@ -71,6 +71,7 @@ func TestDockerCLIBuild(t *testing.T) {
 		testutil.Run(t, test.description, func(t *testutil.T) {
 			t.NewTempDir().Chdir()
 			dockerfilePath, _ := filepath.Abs("Dockerfile")
+			t.Override(&docker.DefaultAuthHelper, testAuthHelper{})
```
Comment (tejal29) [inline pkg/skaffold/build/local/docker_test.go]: i am guessing this is unrelated to this change.

**Comment by dgageot** [comment]
@tejal29 @nkubala this should be better now

File: pkg/skaffold/docker/image.go (hunk 3 above)
Comment (dgageot) [reply pkg/skaffold/docker/image.go]: done

File: pkg/skaffold/docker/image.go (hunk 2 above)
Comment (dgageot) [reply pkg/skaffold/docker/image.go]: done
